File: src/test/java/com/runninglane/dto/buddy/test/java/cases/nonpublic/DtoWithProtectedGetterTest.java
package com.runninglane.dto.buddy.test.java.cases.nonpublic;

import com.runninglane.dto.buddy.exception.DtoBuddyBadInputException;
import com.runninglane.dto.buddy.javainterop.DtoBuddy;
import com.runninglane.dto.buddy.javainterop.bytecode.ThreeStepsByteCodeStrategy;
import com.runninglane.dto.buddy.javainterop.bytecode.compile.CompileJavaByteCodeStrategyCompliment;
import com.squareup.javapoet.MethodSpec;
import com.squareup.javapoet.TypeSpec;
import org.jetbrains.annotations.NotNull;
import org.junit.jupiter.api.Test;

import javax.lang.model.element.Modifier;
import java.lang.reflect.Method;
import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertThrows;

public class DtoWithProtectedGetterTest {

    public abstract static class TestDtoWithProtectedAbstractGetter {
        protected abstract String getName();
    }

    @Test
    public void shouldFailForTestDtoWithProtectedAbstractGetter() {
        DtoBuddy dtoBuddy = new DtoBuddy();
        assertThrows(DtoBuddyBadInputException.class, () -> dtoBuddy.implement(TestDtoWithProtectedAbstractGetter.class));
    }

    @Test
    public void shouldSucceedForTestDtoWithProtectedAbstractGetterWhenCustomizationApplied() {
        class TestByteCodeStrategyCompliment extends CompileJavaByteCodeStrategyCompliment {
            public @NotNull TypeSpec.Builder handleOtherAbstractMethods(
                @NotNull TypeSpec.Builder builder,
                @NotNull List<Method> methods,
                Map<String, ? extends Class<?>> typeParamsMapByName,
                Object dataCollector
            ) {
                TypeSpec.Builder updatedBuilder = builder.addMethod(
                    MethodSpec.methodBuilder("getName")
                        .addModifiers(Modifier.PROTECTED)
                        .addAnnotation(Override.class)
                        .returns(String.class)
                        .addCode("return \"Hello\";\n")
                        .build()
                );
                List<Method> unhandledMethods = methods.stream()
                    .filter(f -> !f.getName().equals("getName"))
                    .toList();
                return super.handleOtherAbstractMethods(updatedBuilder, unhandledMethods, typeParamsMapByName, dataCollector);
            }
        }

        ThreeStepsByteCodeStrategy<?> byteCodeStrategy = new ThreeStepsByteCodeStrategy<>(new TestByteCodeStrategyCompliment());
        DtoBuddy dtoBuddy = new DtoBuddy(byteCodeStrategy);
        dtoBuddy.implement(TestDtoWithProtectedAbstractGetter.class);
    }
}

File: src/test/kotlin/com/runninglane/dto/buddy/test/cases/nonpublic/DtoWithProtectedGetterTest.kt
package com.runninglane.dto.buddy.test.cases.nonpublic

import com.runninglane.dto.buddy.DtoBuddy
import com.runninglane.dto.buddy.bytecode.codegen.CodeGenBasedByteCodeStrategy
import com.runninglane.dto.buddy.bytecode.codegen.KotlinCodeCompiler
import com.runninglane.dto.buddy.bytecode.codegen.KotlinCodeGenerator
import com.runninglane.dto.buddy.exception.DtoBuddyBadInputException
import com.squareup.kotlinpoet.FunSpec
import com.squareup.kotlinpoet.PropertySpec
import com.squareup.kotlinpoet.TypeSpec
import org.junit.jupiter.api.assertThrows
import kotlin.reflect.KClass
import kotlin.reflect.KFunction
import kotlin.reflect.KProperty
import kotlin.test.Test

class DtoWithProtectedGetterTest {

    abstract class TestDtoWithProtectedAbstractImmutableProperty {
        protected abstract val name: String
    }

    abstract class TestDtoWithProtectedAbstractGetter {
        protected abstract fun getName(): String
    }

    @Test
    fun shouldFailForTestDtoWithProtectedAbstractImmutableProperty() {
        val dtoBuddy = DtoBuddy()
        assertThrows<DtoBuddyBadInputException> {
            dtoBuddy.implement(TestDtoWithProtectedAbstractImmutableProperty::class)
        }
    }

    @Test
    fun shouldFailForTestDtoWithProtectedAbstractGetter() {
        val dtoBuddy = DtoBuddy()
        assertThrows< DtoBuddyBadInputException> {
            dtoBuddy.implement(TestDtoWithProtectedAbstractGetter::class)
        }
    }

    @Test
    fun shouldSucceedForTestDtoWithProtectedConcreteImmutablePropertyWhenCustomizationApplied() {
        class TestCodeGenerator : KotlinCodeGenerator() {
            override fun handleOtherAbstractMembers(
                builder: TypeSpec.Builder,
                properties: List<KProperty<*>>,
                functions: List<KFunction<*>>,
                typeParamsMapByName: Map<String, KClass<*>>?,
                dataCollector: Any?
            ): TypeSpec.Builder {
                val updatedBuilder = builder.addProperty(
                    PropertySpec.builder("name", String::class)
                        .addModifiers(com.squareup.kotlinpoet.KModifier.OVERRIDE)
                        .initializer("\"Hello\"")
                        .build()
                )
                val unhandledProperties = properties.filter { it.name != "name" }
                return super.handleOtherAbstractMembers(updatedBuilder, unhandledProperties, functions, typeParamsMapByName, dataCollector)
            }
        }

        val byteCodeStrategy = CodeGenBasedByteCodeStrategy(TestCodeGenerator(), KotlinCodeCompiler())
        val dtoBuddy = DtoBuddy(byteCodeStrategy)
        dtoBuddy.implement(TestDtoWithProtectedAbstractImmutableProperty::class)
    }

    @Test
    fun shouldSucceedForTestDtoWithProtectedAbstractGetterWhenCustomizationApplied() {
        class TestCodeGenerator : KotlinCodeGenerator() {
            override fun handleOtherAbstractMembers(
                builder: TypeSpec.Builder,
                properties: List<KProperty<*>>,
                functions: List<KFunction<*>>,
                typeParamsMapByName: Map<String, KClass<*>>?,
                dataCollector: Any?
            ): TypeSpec.Builder {
                val updatedBuilder = builder.addFunction(
                    FunSpec.builder("getName")
                        .addModifiers(com.squareup.kotlinpoet.KModifier.OVERRIDE)
                        .returns(String::class)
                        .addStatement("return \"Hello\"")
                        .build()
                )
                val unhandledFunctions = functions.filter { it.name != "getName" }
                return super.handleOtherAbstractMembers(updatedBuilder, properties, unhandledFunctions, typeParamsMapByName, dataCollector)
            }
        }

        val byteCodeStrategy = CodeGenBasedByteCodeStrategy(TestCodeGenerator(), KotlinCodeCompiler())
        val dtoBuddy = DtoBuddy(byteCodeStrategy)
        dtoBuddy.implement(TestDtoWithProtectedAbstractGetter::class)
    }
}